File: LeetCode/875.Koko_Eating_Bananas/875.Koko_Eating_Bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        // left = 1, right = max(piles)
        int left = 1, right = -1;
        for (int p : piles) {
            right = Math.max(right, p);
        }

        // binary search
        while (left < right) {
            int mid = (left + right) / 2;
            long hours = getHours(piles, mid);
            // System.out.println("hours:" + hours + " mid:" + mid + " left:" + left + "
            // right:" + right);
            if (hours <= h) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        // left = right
        // while (getHours(piles, left) > h) {
        // left += 1;
        // }

        return left;
    }

    private long getHours(int[] piles, int k) {
        long hours = 0;
        for (int p : piles) {
            hours += (p / k);
            if (p % k > 0)
                hours += 1;
        }
        return hours;
    }
}

File: LeetCode/875.Koko_Eating_Bananas/875.Koko_Eating_Bananas.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:

        def check(k: int) -> bool:
            res = 0
            for p in piles:
                res += (p + k - 1) // k
                if res > h:
                    return False
            return res <= h

        left, right = 1, max(piles) + 1
        while left < right:
            mid = (left + right) // 2
            if check(mid):
                right = mid
            else:
                left = mid + 1

        return left
